.grit/patterns/hathora_to_new_ts_sdk.md

```diff
@@ -0,0 +1,226 @@
+  } else if ($resource <: `AuthV1Api`) {
+    $result = `authV1`
+  } else if ($resource <: `LobbyV1Api`) {
+    $result = `lobbyV1`
+  } else if ($resource <: `LobbyV2Api`) {
+    $result = `lobbyV2`
+  },
+  return $result
+}
+
+// some functions have changed names to reflect deprecation.
+// XXX not all of them, eg lobby v2 `createLocalLobby` is deprecated,
+// but not renamed.
+function deprecated_suffix($resource, $method) {
+  if ($resource <: or {`RoomV1Api`, `LobbyV1Api`}) {
+    return `Deprecated`
+  },
+  return ``
+}
+
+// TODO the rest of the unwrappings
+function wrapper_key($method) {
+  if ($method <: or {`loginAnonymous`, `loginGoogle`}) {
+    return `loginResponse`
+  } else if ($method <: `getLobbyInfo`) {
+    return `lobby`
+  } else {
+    return `connectionInfo`
+  },
+  return `connectionInfo`
+}
+
+// many request funs have a different arg order, usually shifting the request to the beginning,
+// so we need to rearrange $args
+// TODO: the rest of the functions
+function reorder_args($method, $args) {
+  if ($method <: `setLobbyState`) {
+    // TODO how do we match and use rest-of-args without so much copypasta?
+    if ($args <: [$app_id, $room_id, $req, $opts]) {
+      $use_args = join(list=[$req, $app_id, $room_id, $opts], separator=", ")
+    } else if ($args <: [$app_id, $room_id, $req]) {
+      $use_args = join(list=[$req, $room_id, $app_id], separator=", ")
+    } else {
+      $use_args = $args
+    }
+  } else if ($method <: `loginGoogle`) {
+    if ($args <: [$app_id, $req, $opts]) {
+      $use_args = join(list=[$req, $app_id, $opts], separator=", ")
+    } else if ($args <: [$app_id, $req]) {
+      $use_args = join(list=[$req, $app_id], separator=", ")
+    } else {
+      $use_args = $args
+    }
+  } else {
+    $use_args = join(list=$args, separator=", ")
+  },
+  return $use_args
+}
+
+pattern rewrite_method_calls($resource, $callee) {
+  maybe any {
+    // TODO more elegant handling of await-or-not, maybe via regex? better idea?
+    contains bubble($resource, $callee) `await $callee.$method($args)` as $body where {
+      $suffix = deprecated_suffix(resource=$resource, method=$method),
+      // TODO finish rearranging methods
+      // TODO generalize arg-rearrangement for the non-await case, since they're the same.
+      $use_args = reorder_args($method, $args),
+      $unwrap_at = wrapper_key(method=$method),
+      $body => `(await $callee.$[method]$[suffix]($use_args)).$unwrap_at`
+    },
+    contains bubble($resource, $callee) `$callee.$method($args)` as $body where {
+      $suffix = deprecated_suffix(resource=$resource, method=$method),
+      $use_args = reorder_args($method, $args),
+      $body => `$callee.$[method]$[suffix]($use_args)`
+    }
+  }
+}
+
+pattern generic_replace_resource($resource) {
+  and {
+      replace_resource_import($resource),
+    // TODO what do we do about hygiene when we import names like `sdk`?
+    // do we have to worry about shadowing?
```

Comment:
Not for now, but you can also easily add a side condition to check if `sdk` is already used and add a suffix if necessary.